File: neel/C_All_in_one_Gun.cpp
#include <bits/stdc++.h>
using namespace std;

using ll = long long;

void solve() {
    ll n, h, k;
    cin >> n >> h >> k;

    vector<ll> a(n);
    for (ll &x : a) cin >> x;

    ll s = 0;
    for (ll x : a) s += x;

    ll x = h / s;
    ll y = h % s;

    ll ans = x * n + max(0LL, x - 1) * k;

    if (y > 0) ans += k;

    vector<ll> sff = a;

    for (ll i = n - 2; i >= 0; i--)
        sff[i] = max(sff[i], sff[i + 1]);

    for (ll i = 0; i < n; i++) {
        if (sff[i] > y) {
            ans += 1;
            break;
        } else {
            y -= a[i];
            ans += 1;
        }
    }

    cout << ans - 1 << '\n';
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int t;
    cin >> t;
    while (t--) solve();

    return 0;
}

File: neel/C_All_in_one_Gun.py
PI = 3.14159265358979323846
E = 2.718281828459045
INF = 1000000000000000000
MOD1 = 1000000007
MOD2 = 998244353

def solve():
    n,h,k = map(int,input().split())
    a = list(map(int,input().split()))
    s = sum(a)

    x = h//s
    y = h%s

    ans = x*n + max(0,x-1)*k

    if y > 0:
        if x > 0:
            ans += k

        sff = a[:]

        for i in range(n-2,-1,-1):
            sff[i] = max(sff[i], sff[i+1])

        ps = 0
        mn = 10**18

        for i in range(n):
            ps += a[i]
            mn = min(mn, a[i])

            if ps >= y:
                ans += i+1
                break

            if i < n-1:
                new_ps = ps - mn + sff[i+1]
                if new_ps >= y:
                    ans += i+1
                    break

    print(ans)

def main():
    t = 1
    t = int(input()) 
    for _ in range(t):
        solve()

main()